File: BackendAppMusic/src/controllers/authController.ts
import { Request, Response } from 'express';
import { AuthService } from '../services/authService';
import { Auth, ApiResponse, ErrorCode } from '../types/api.types';
import { Pool } from 'pg';

export class AuthController {
    private authService: AuthService;

    constructor(pool: Pool) {
        this.authService = new AuthService(
            pool,
            {
                jwtSecret: process.env.JWT_SECRET || 'your-super-secret-jwt-key',
                jwtRefreshSecret: process.env.JWT_REFRESH_SECRET || 'your-super-secret-refresh-key',
                accessTokenExpiry: 15, // 15 minutes
                refreshTokenExpiry: 7, // 7 days
                emailFrom: process.env.EMAIL_FROM || 'noreply@appmusic.com',
                emailService: null, // Configure nodemailer here
            }
        );
    }

    /**
     * Register user
     * POST /api/v1/auth/register
     */
    async register(req: Request, res: Response): Promise<void> {
        try {
            const { email, password, name } = req.body;

            // Validate required fields
            if (!email || !password || !name) {
                res.status(400).json({
                    success: false,
                    code: ErrorCode.VALIDATION_ERROR,
                    message: 'Email, password, and name are required',
                    statusCode: 400,
                } as ApiResponse);
                return;
            }

            const result = await this.authService.register({
                email,
                password,
                name,
            });

            res.status(201).json({
                success: true,
                data: result,
                message: 'User registered successfully',
                statusCode: 201,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Login user
     * POST /api/v1/auth/login
     */
    async login(req: Request, res: Response): Promise<void> {
        try {
            const { email, password } = req.body;

            if (!email || !password) {
                res.status(400).json({
                    success: false,
                    code: ErrorCode.VALIDATION_ERROR,
                    message: 'Email and password are required',
                    statusCode: 400,
                } as ApiResponse);
                return;
            }

            const ipAddress = req.ip || 'unknown';
            const userAgent = req.get('user-agent') || 'unknown';

            const result = await this.authService.login(
                { email, password },
                ipAddress,
                userAgent
            );

            res.status(200).json({
                success: true,
                data: result,
                message: 'Login successful',
                statusCode: 200,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Refresh token
     * POST /api/v1/auth/refresh-token
     */
    async refreshToken(req: Request, res: Response): Promise<void> {
        try {
            const { refreshToken } = req.body;

            if (!refreshToken) {
                res.status(400).json({
                    success: false,
                    code: ErrorCode.VALIDATION_ERROR,
                    message: 'Refresh token is required',
                    statusCode: 400,
                } as ApiResponse);
                return;
            }

            const result = await this.authService.refreshToken(refreshToken);

            res.status(200).json({
                success: true,
                data: result,
                message: 'Token refreshed successfully',
                statusCode: 200,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Logout user
     * POST /api/v1/auth/logout
     */
    async logout(req: Request, res: Response): Promise<void> {
        try {
            const userId = (req as any).userId;
            const { refreshToken } = req.body;

            if (!userId) {
                res.status(401).json({
                    success: false,
                    code: ErrorCode.UNAUTHORIZED,
                    message: 'Unauthorized',
                    statusCode: 401,
                } as ApiResponse);
                return;
            }

            await this.authService.logout(userId, refreshToken);

            res.status(200).json({
                success: true,
                message: 'Logged out successfully',
                statusCode: 200,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Verify email
     * GET /api/v1/auth/verify-email/:token
     */
    async verifyEmail(req: Request, res: Response): Promise<void> {
        try {
            const { token } = req.params;

            if (!token) {
                res.status(400).json({
                    success: false,
                    code: ErrorCode.VALIDATION_ERROR,
                    message: 'Verification token is required',
                    statusCode: 400,
                } as ApiResponse);
                return;
            }

            await this.authService.verifyEmail(token);

            res.status(200).json({
                success: true,
                message: 'Email verified successfully',
                statusCode: 200,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Request password reset
     * POST /api/v1/auth/request-password-reset
     */
    async requestPasswordReset(req: Request, res: Response): Promise<void> {
        try {
            const { email } = req.body;

            if (!email) {
                res.status(400).json({
                    success: false,
                    code: ErrorCode.VALIDATION_ERROR,
                    message: 'Email is required',
                    statusCode: 400,
                } as ApiResponse);
                return;
            }

            await this.authService.requestPasswordReset(email);

            // Always return success to prevent email enumeration
            res.status(200).json({
                success: true,
                message: 'If the email exists, a password reset link has been sent',
                statusCode: 200,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Reset password
     * POST /api/v1/auth/reset-password
     */
    async resetPassword(req: Request, res: Response): Promise<void> {
        try {
            const { token, newPassword } = req.body;

            if (!token || !newPassword) {
                res.status(400).json({
                    success: false,
                    code: ErrorCode.VALIDATION_ERROR,
                    message: 'Token and new password are required',
                    statusCode: 400,
                } as ApiResponse);
                return;
            }

            await this.authService.resetPassword(token, newPassword);

            res.status(200).json({
                success: true,
                message: 'Password reset successfully',
                statusCode: 200,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Change password (authenticated)
     * POST /api/v1/auth/change-password
     */
    async changePassword(req: Request, res: Response): Promise<void> {
        try {
            const userId = (req as any).userId;
            const { currentPassword, newPassword } = req.body;

            if (!userId) {
                res.status(401).json({
                    success: false,
                    code: ErrorCode.UNAUTHORIZED,
                    message: 'Unauthorized',
                    statusCode: 401,
                } as ApiResponse);
                return;
            }

            if (!currentPassword || !newPassword) {
                res.status(400).json({
                    success: false,
                    code: ErrorCode.VALIDATION_ERROR,
                    message: 'Current password and new password are required',
                    statusCode: 400,
                } as ApiResponse);
                return;
            }

            await this.authService.changePassword(userId, currentPassword, newPassword);

            res.status(200).json({
                success: true,
                message: 'Password changed successfully',
                statusCode: 200,
            } as ApiResponse);
        } catch (error: any) {
            this.handleError(error, res);
        }
    }

    /**
     * Handle errors and send appropriate response
     */
    private handleError(error: any, res: Response): void {
        const code = error.code || ErrorCode.INTERNAL_ERROR;
        const statusCode = error.statusCode || 500;
        const message = error.message || 'Internal server error';

        res.status(statusCode).json({
            success: false,
            code,
            message,
            statusCode,
            errors: error.errors || [],
        } as ApiResponse);
    }
}

File: AppMusic/melody/lib/provider/auth_controller.dart
import 'package:get/get.dart';
import '../models/auth_models.dart';
import '../services/auth_service.dart';
import '../services/token_storage_service.dart';

class AuthController extends GetxController {
  final AuthService _authService;

  AuthController({
    required AuthService authService,
    required TokenStorageService tokenStorage,
  }) : _authService = authService;

  // Reactive variables
  final Rx<User?> user = Rx<User?>(null);
  final RxBool isLoggedIn = RxBool(false);
  final RxBool isLoading = RxBool(false);
  final RxString errorMessage = RxString('');
  final RxBool isEmailVerified = RxBool(false);
  final Rx<String?> accessToken = Rx<String?>(null);

  @override
  void onInit() {
    super.onInit();
    _checkAuthStatus();
  }

  /// Check if user is already logged in
  Future<void> _checkAuthStatus() async {
    try {
      final isAuth = await _authService.isAuthenticated();
      if (isAuth) {
        final storedUser = await _authService.getStoredUser();
        if (storedUser != null) {
          user.value = storedUser;
          isLoggedIn.value = true;
          isEmailVerified.value = storedUser.isEmailVerified ?? false;

          final token = await _authService.getAccessToken();
          accessToken.value = token;
        }
      }
    } catch (e) {
      print('Error checking auth status: $e');
    }
  }

  /// Register new user
  Future<void> register({
    required String email,
    required String password,
    required String name,
  }) async {
    try {
      isLoading.value = true;
      errorMessage.value = '';

      final response = await _authService.register(
        email: email,
        password: password,
        name: name,
      );

      if (response.user != null) {
        user.value = response.user;
        isLoggedIn.value = true;
        isEmailVerified.value = response.user!.isEmailVerified ?? false;

        final token = await _authService.getAccessToken();
        accessToken.value = token;

        Get.offAllNamed('/home');
      } else {
        errorMessage.value = response.message ?? 'Registration failed';
      }
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      Get.snackbar(
        'Registration Error',
        errorMessage.value,
        duration: Duration(seconds: 3),
      );
    } finally {
      isLoading.value = false;
    }
  }

  /// Login user
  Future<void> login({
    required String email,
    required String password,
  }) async {
    try {
      isLoading.value = true;
      errorMessage.value = '';

      final response = await _authService.login(
        email: email,
        password: password,
      );

      if (response.user != null) {
        user.value = response.user;
        isLoggedIn.value = true;
        isEmailVerified.value = response.user!.isEmailVerified ?? false;

        final token = await _authService.getAccessToken();
        accessToken.value = token;

        Get.offAllNamed('/home');
      } else {
        errorMessage.value = response.message ?? 'Login failed';
      }
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      Get.snackbar(
        'Login Error',
        errorMessage.value,
        duration: Duration(seconds: 3),
      );
    } finally {
      isLoading.value = false;
    }
  }

  /// Logout user
  Future<void> logout() async {
    try {
      isLoading.value = true;

      await _authService.logout();

      user.value = null;
      isLoggedIn.value = false;
      isEmailVerified.value = false;
      accessToken.value = null;
      errorMessage.value = '';

      Get.offAllNamed('/login');
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      Get.snackbar(
        'Logout Error',
        errorMessage.value,
        duration: Duration(seconds: 3),
      );
    } finally {
      isLoading.value = false;
    }
  }

  /// Request password reset
  Future<void> requestPasswordReset(String email) async {
    try {
      isLoading.value = true;
      errorMessage.value = '';

      final response = await _authService.requestPasswordReset(email);

      if (response.success) {
        Get.snackbar(
          'Success',
          'Password reset link sent to your email',
          duration: Duration(seconds: 3),
        );
        Get.toNamed('/login');
      } else {
        errorMessage.value = response.message ?? 'Failed to send reset link';
        Get.snackbar(
          'Error',
          errorMessage.value,
          duration: Duration(seconds: 3),
        );
      }
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      Get.snackbar(
        'Error',
        errorMessage.value,
        duration: Duration(seconds: 3),
      );
    } finally {
      isLoading.value = false;
    }
  }

  /// Reset password with token
  Future<void> resetPassword({
    required String token,
    required String newPassword,
  }) async {
    try {
      isLoading.value = true;
      errorMessage.value = '';

      final response = await _authService.resetPassword(
        token: token,
        newPassword: newPassword,
      );

      if (response.success) {
        Get.snackbar(
          'Success',
          'Password reset successfully',
          duration: Duration(seconds: 3),
        );
        Get.offAllNamed('/login');
      } else {
        errorMessage.value = response.message ?? 'Failed to reset password';
        Get.snackbar(
          'Error',
          errorMessage.value,
          duration: Duration(seconds: 3),
        );
      }
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      Get.snackbar(
        'Error',
        errorMessage.value,
        duration: Duration(seconds: 3),
      );
    } finally {
      isLoading.value = false;
    }
  }

  /// Verify email
  Future<void> verifyEmail(String token) async {
    try {
      isLoading.value = true;
      errorMessage.value = '';

      final response = await _authService.verifyEmail(token);

      if (response.success) {
        isEmailVerified.value = true;
        Get.snackbar(
          'Success',
          'Email verified successfully',
          duration: Duration(seconds: 3),
        );
        Get.toNamed('/home');
      } else {
        errorMessage.value = response.message ?? 'Email verification failed';
        Get.snackbar(
          'Error',
          errorMessage.value,
          duration: Duration(seconds: 3),
        );
      }
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      Get.snackbar(
        'Error',
        errorMessage.value,
        duration: Duration(seconds: 3),
      );
    } finally {
      isLoading.value = false;
    }
  }

  /// Get current user profile
  Future<void> getCurrentUser() async {
    try {
      isLoading.value = true;

      final currentUser = await _authService.getCurrentUser();
      user.value = currentUser;
      isEmailVerified.value = currentUser.isEmailVerified ?? false;
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      print('Error fetching user: $errorMessage');
    } finally {
      isLoading.value = false;
    }
  }

  /// Change password
  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    try {
      isLoading.value = true;
      errorMessage.value = '';

      final response = await _authService.changePassword(
        currentPassword: currentPassword,
        newPassword: newPassword,
      );

      if (response.success) {
        Get.snackbar(
          'Success',
          'Password changed successfully',
          duration: Duration(seconds: 3),
        );
        Get.back();
      } else {
        errorMessage.value = response.message ?? 'Failed to change password';
        Get.snackbar(
          'Error',
          errorMessage.value,
          duration: Duration(seconds: 3),
        );
      }
    } catch (e) {
      errorMessage.value = e.toString().replaceAll('Exception: ', '');
      Get.snackbar(
        'Error',
        errorMessage.value,
        duration: Duration(seconds: 3),
      );
    } finally {
      isLoading.value = false;
    }
  }

  /// Refresh access token (called by interceptor)
  Future<bool> refreshAccessToken() async {
    try {
      final response = await _authService.refreshToken();
      accessToken.value = response.accessToken;
      return true;
    } catch (e) {
      print('Token refresh failed: $e');
      // If refresh fails, logout user
      await logout();
      return false;
    }
  }

  /// Clear error message
  void clearError() {
    errorMessage.value = '';
  }

  /// Get user display name
  String get displayName => user.value?.name ?? 'User';

  /// Get user email
  String get userEmail => user.value?.email ?? '';

  /// Check if user is premium
  bool get isPremium => user.value?.isPremium ?? false;

  /// Check if user profile is complete
  bool get isProfileComplete {
    return user.value != null &&
        user.value!.email.isNotEmpty &&
        user.value!.name.isNotEmpty;
  }
}
